### PR TITLE
[ConstraintSystem] Remove an unsound optimization hack for generic fu…

### PR DESCRIPTION
**Pull request by rudkx**
…nctions.

I think there is something far more narrow we could do here, but I'm
not sure if there is real code that such a hack would benefit from. We
can leave that to consider another day.

Fixes rdar://problem/40819547 (aka https://bugs.swift.org/browse/SR-7884).

### CODE REVIEW

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

File: test/Constraints/sr7884.swift [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// SR-7884
+func f<T>(_ x: T) -> T {
+  return x
+}
+
+func f<T>(_ x: T?) -> T? {
+  return x
+}
+
+let r = f(1) // calls optional variant
```
Comment (slavapestov) [inline test/Constraints/sr7884.swift]: shouldn't this read "calls non-optional variant"?

File: validation-test/Sema/type_checker_perf/slow/more_specialized_generic_func.swift.gyb [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// RUN: %scale-test --begin 1 --end 10 --step 1 --select incrementScopeCounter %s --expected-exit-code 0
+// RUN: %scale-test --invert-result --begin 1 --end 10 --step 1 --select incrementScopeCounter %s --expected-exit-code 0
```
Comment (slavapestov) [inline validation-test/Sema/type_checker_perf/slow/more_specialized_generic_func.swift.gyb]: Oh well...

File: test/Constraints/sr7884.swift (hunk 1 above)
Comment (rudkx) [reply test/Constraints/sr7884.swift]: Yeah, that was a holdover from the original test case. I'll remove the comment.

**Comment by rudkx** [comment]
The source compatibility failure appears to be due to an SDK change in Xcode 10.0 beta 1.

**Comment by DougGregor** [comment]
We should do something to narrow the performance hack, not revert it.

**Comment by DougGregor** [comment]
Actually, I'm wondering if maybe we should strip top-level optionals when comparing overloads.

**Comment by xedin** [comment]
I've looked at how we compare declarations again and it seems like this optional check should be the other way https://github.com/apple/swift/blob/master/lib/Sema/CSRanking.cpp#L708 because we replace parameters in reverse, so if the score got increased like that first one is actually better. WDYT?

**Pull request merged**